numpy/core/einsumfunc.pyi

```diff
@@ -0,0 +1,138 @@
+import sys
+from typing import List, TypeVar, Optional, Any, overload, Union, Tuple, Sequence
+
+from numpy import (
+    ndarray,
+    dtype,
+    bool_,
+    unsignedinteger,
+    signedinteger,
+    floating,
+    complexfloating,
+    number,
+    _OrderKACF,
+)
+from numpy.typing import (
+    _ArrayOrScalar,
+    _ArrayLikeBool_co,
+    _ArrayLikeUInt_co,
+    _ArrayLikeInt_co,
+    _ArrayLikeFloat_co,
+    _ArrayLikeComplex_co,
+    _DTypeLikeBool,
+    _DTypeLikeUInt,
+    _DTypeLikeInt,
+    _DTypeLikeFloat,
+    _DTypeLikeComplex,
+    _DTypeLikeComplex_co,
+)
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+_ArrayType = TypeVar(
+    "_ArrayType",
+    bound=ndarray[Any, dtype[Union[bool_, number[Any]]]],
+)
+
+_OptimizeKind = Union[
+    None, bool, Literal["greedy", "optimal"], Sequence[Any]
```

Comment:
Typing of the explicit contraction list (see [`np.einsum_path`](https://numpy.org/doc/stable/reference/generated/numpy.einsum_path.html)) is beyond what we can currently do unfortunately, _i.e._ a sequence that starts with a string followed by a variable number of integer tuples.

For now it's type is simply set to `Sequence[Any]`.